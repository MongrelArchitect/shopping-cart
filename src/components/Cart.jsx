import React from 'react';

export default function Cart({ cart, products, removeFromCart }) {
  const getTotalPrice = () => {
    const keys = Object.keys(cart);
    let total = 0;
    keys.forEach((key) => {
      total += products[key - 1].price * cart[key];
    });
    return Math.round((total + Number.EPSILON) * 100) / 100;
  };

  const handleRemove = (event) => {
    removeFromCart(event.target.dataset.key);
  };

  const cartKeys = Object.keys(cart);

  if (!cartKeys.length) {
    return (
      <div className="cart">
        <h3>Cart is empty. Get shopping!</h3>
      </div>
    );
  }

  return (
    <div className="cart">
      <ul>
        {cartKeys.map((key) => {
          const id = key - 1;
          return (
            <li key={key}>
              <img
                src={products[id].image}
                alt=""
                className="cart-product-image"
              />
              <span>
                $
                {products[id].price}
                {' x '}
                {cart[key]}
                : $
                {products[id].price * cart[key]}
                <button
                  className="remove"
                  data-key={key}
                  type="button"
                  onClick={handleRemove}
                >
                  âœ•
                </button>
              </span>
            </li>
          );
        })}
        <li>
          <span />
          <span className="cart-total">
            Total: $
            {getTotalPrice()}
            <button type="button">CHECKOUT</button>
          </span>
        </li>
      </ul>
    </div>
  );
}
