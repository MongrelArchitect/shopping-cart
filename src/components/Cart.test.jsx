import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Cart from './Cart';

describe('Cart component', () => {
  const mockCart = {
    1: 7,
    3: 2,
    4: 3,
  };

  const mockProducts = [
    {
      image: '1',
      price: 100,
    },
    {
      image: '2',
      price: 10,
    },
    {
      image: '3',
      price: 250,
    },
    {
      image: '4',
      price: 1000,
    },
  ];

  const mockRemove = jest.fn();

  it('renders correctly with zero items in cart', () => {
    const { container } = render(<Cart cart={{}} />);
    expect(container).toMatchSnapshot();
  });

  it('renders correctly with some items in cart', () => {
    const { container } = render(
      <Cart cart={mockCart} products={mockProducts} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('calculates cart total correctly', () => {
    render(<Cart cart={mockCart} products={mockProducts} />);
    const total = screen.getByText(/Total/);
    expect(total.textContent).toContain('Total: $4200');
  });

  it('displays correct number of "delete" buttons', () => {
    render(
      <Cart
        cart={mockCart}
        products={mockProducts}
        removeFromCart={mockRemove}
      />,
    );
    const buttons = screen.getAllByText('✕');
    expect(buttons.length).toBe(3);
  });

  it('calls removeFromCart correct number of times', () => {
    render(
      <Cart
        cart={mockCart}
        products={mockProducts}
        removeFromCart={mockRemove}
      />,
    );
    const buttons = screen.getAllByText('✕');
    userEvent.click(buttons[0]);
    expect(mockRemove).toHaveBeenCalledTimes(1);
    userEvent.click(buttons[1]);
    expect(mockRemove).toHaveBeenCalledTimes(2);
  });

  it('calls removeFromCart with correct arguments', () => {
    render(
      <Cart
        cart={mockCart}
        products={mockProducts}
        removeFromCart={mockRemove}
      />,
    );
    const buttons = screen.getAllByText('✕');
    userEvent.click(buttons[0]);
    expect(mockRemove).toHaveBeenCalledWith('1');
    userEvent.click(buttons[1]);
    expect(mockRemove).toHaveBeenCalledWith('3');
    userEvent.click(buttons[2]);
    expect(mockRemove).toHaveBeenCalledWith('4');
  });
});
