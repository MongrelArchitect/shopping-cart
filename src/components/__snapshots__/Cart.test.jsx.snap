// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component renders correctly with some items in cart 1`] = `
<div>
  <div
    class="cart"
  >
    <ul>
      <li>
        <img
          alt=""
          class="cart-product-image"
          src="1"
        />
        <span>
          $
          100
           x 
          7
          : $
          700
          <button
            class="remove"
            data-key="1"
            type="button"
          >
            ✕
          </button>
        </span>
      </li>
      <li>
        <img
          alt=""
          class="cart-product-image"
          src="3"
        />
        <span>
          $
          250
           x 
          2
          : $
          500
          <button
            class="remove"
            data-key="3"
            type="button"
          >
            ✕
          </button>
        </span>
      </li>
      <li>
        <img
          alt=""
          class="cart-product-image"
          src="4"
        />
        <span>
          $
          1000
           x 
          3
          : $
          3000
          <button
            class="remove"
            data-key="4"
            type="button"
          >
            ✕
          </button>
        </span>
      </li>
      <li>
        <span />
        <span
          class="cart-total"
        >
          Total: $
          4200
          <button
            type="button"
          >
            CHECKOUT
          </button>
        </span>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Cart component renders correctly with zero items in cart 1`] = `
<div>
  <div
    class="cart"
  >
    <h3>
      Cart is empty. Get shopping!
    </h3>
  </div>
</div>
`;
